File: 1189-maximum-number-of-balloons/1189-maximum-number-of-balloons.cpp
class Solution {
public:
    int maxNumberOfBalloons(string text) {
        int n = text.size();
        unordered_map<char, int> mp;
        unordered_set<char> set;
        string s = "balloon";
        for(int i = 0; i < s.length(); i++) {
            set.insert(s[i]);
        }
        for(int i = 0; i < n; i++) {
            if(set.find(text[i]) != set.end())
                mp[text[i]]++;
        }

        int ans = 0;
        int m = text.length()/s.length();
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < s.length(); j++) {
                if(mp.find(s[j]) != mp.end()) {
                    mp[s[j]]--;
                    if(mp[s[j]] == 0) 
                        mp.erase(s[j]);
                }
                else return ans;
            }
            ans++;
        }

        return ans;
    }
};

File: 1189-maximum-number-of-balloons/1189-maximum-number-of-balloons.py
class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        n = len(text)
        mp = {}
        sett = {'b', 'a', 'l', 'o', 'n'}

        for i in range(n):
            if text[i] in sett:
                mp[text[i]] = mp.get(text[i], 0) + 1
        
        s = "balloon"
        m = len(text)//len(s)

        ans = 0
        for i in range(m):
            for j in range(len(s)):
                if s[j] in mp:
                    mp[s[j]] -= 1
                    if(mp[s[j]] == 0):
                        del mp[s[j]]
                else:
                    return ans
            ans += 1

        return ans